src/index.js

```diff
@@ -609,12 +610,40 @@ const members = {
       const getHtmlNodeOuterHtml = (node) => node.outerHTML;
       const getOuterHtmlByNodeType = {
         1: getHtmlNodeOuterHtml,
-        3: getEscapedTextContent,
         8: getCommentNodeOuterHtml,
       };
+      // Text nodes with these ancestors should be treated as raw text
+      // See section 8.4 of
+      // https://www.w3.org/TR/2008/WD-html5-20080610/serializing.html#html-fragment
+      const rawTextNodeAncestors = {
+        style: true,
+        script: true,
+        xmp: true,
+        iframe: true,
+        noembed: true,
+        noframes: true,
+        noscript: true,
+        plaintext: true,
+      };
 
       eachChildNode(this, node => {
-        const getOuterHtml = getOuterHtmlByNodeType[node.nodeType] || getHtmlNodeOuterHtml;
+        let getOuterHtml;
+        if (node.nodeType === 3) {
+          let parentNode = node.parentNode;
+          while (parentNode) {
+            if (parentNode.nodeType === 1 && // only for elements
+                parentNode.nodeName.toLowerCase() in rawTextNodeAncestors) {
+              getOuterHtml = getRawTextContent;
```

Comment:
In terms of the tags where this behaviour is described:

- `style` - text is the immediate child
- `script` - text is the immediate child
- `iframe` - useful content is attributes (src, etc)
- **`noscript`** - can contain HTML
- `xmp` - deprecated
- `noembed` - deprecated
- `noframes` - deprecated
- `plaintext` - deprecated

The only tag that is really impacted is the `noscript` tag, which if we don't look up the tree of a text node, we will encode the innerHTML incorrectly. Perhaps this is something we can live with, as it's probably unlikely to get used too often.